Feature/project fugu (#167)
iodigital-com/io-technology: data/blog/project-fugu.md

---
title: 'Project Fugu'
date: '2023-04-14'
tags: ['frontend', 'web-capabilities', 'project-fugu']
images: ['/articles/project-fugu/header.webp']
summary: Cross-platform development is hard. Each platform has its specific implementation of an API and you end up with separate apps for each platform. Multiple frameworks try to fix this issue by creating an abstract between the platform and the application. Browser vendors are doing the same and it's called Project Fugu.
authors: ['lucien-immink']
theme: 'rouge'
---

> Cross-platform development is hard. Features and capabilities of an application require you to think about how it needs to be implemented in different platforms like Android, iOS, Web, Windows, MacOS and Linux. If you, for example, need to access the address details of a contact that is stored on a device you need to implement a contact picker for all the different platforms you want to support or you need to come up with a completely custom implementation. Custom implementations are probably not what you want to do. It can become quite messy quite fast.

Reviewer: Cross-platform _software_ development?

> Cross-platform development is hard. Features and capabilities of an application require you to think about how it needs to be implemented in different platforms like Android, iOS, Web, Windows, MacOS and Linux. If you, for example, need to access the address details of a contact that is stored on a device you need to implement a contact picker for all the different platforms you want to support or you need to come up with a completely custom implementation. Custom implementations are probably not what you want to do. It can become quite messy quite fast.

Reviewer: to be implemented _on_ different platforms?

> Cross-platform development is hard. Features and capabilities of an application require you to think about how it needs to be implemented in different platforms like Android, iOS, Web, Windows, MacOS and Linux. If you, for example, need to access the address details of a contact that is stored on a device you need to implement a contact picker for all the different platforms you want to support or you need to come up with a completely custom implementation. Custom implementations are probably not what you want to do. It can become quite messy quite fast.

Reviewer: A little bit more information about what a contact picker is?

...to implement a piece of software, that allows you to access and choose a contact, let's call it a contact picker, for all the different platforms...

So what can you do?

> What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, previously known as PhoneGap. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.

Reviewer: ```suggestion
What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, Ionic's Capacitor, Electron or Tauri. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.
```

> What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, Ionic's Capacitor, Electron or Tauri. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.

Reviewer 2: Maybe link the frameworks you're mentioning?

> While technology like Apache Cordova enables developers to build one UI for multiple platforms they still have to build the application for all the platforms they want to support and sometimes the support for a platform or specific feature is a bit flaky.

Reviewer: ```suggestion
While these hybrid frameworks enable developers to build one UI for multiple platforms they still have to build the application for all the platforms they want to support. The support for a platform or platform specific feature can often be flaky.
```

_Often_ because I am speaking from experience 😄

> While these hybrid frameworks enable developers to build one UI for multiple platforms they still have to build the application for all the platforms they want to support. The support for a platform or platform specific feature can often be flaky.

Reviewer 2: Isn't it `platform-specific` instead of `platform specific`?

## Project Fugu

Project Fugu 🐡 is an effort to close gaps in the web's capabilities enabling new classes of applications to run on the web. APIs that Project Fugu is delivering enable new experiences on the web while preserving the web's core benefits of security, low friction, and cross-platform delivery. All Project Fugu API proposals are made in the open and on the standards track.

![Schema of the elements of a URL](/articles/project-fugu/api-call-abstraction.webp)

Fugu is organized as a Chromium project, open to all Chromium contributors and organizations. Today, that includes Microsoft, Intel, Samsung, Electron and Google (among others). Fugu Leads triage incoming requests from partners, determine demand, prioritize them, find champions, track development, and help organize the release and documentation for the capability.

![Schema of the elements of a URL](/articles/project-fugu/adding-apis.webp)

> Let's take a look at some of the features that have already been released:

Reviewer: IMHO, I think this could use an additional h2, right above this sentence.

## Fugu features already available

```html
<form>
  <input autocomplete="one-time-code" required />
  <input type="submit" />
</form>
```

```javascript
if ('OTPCredential' in window) {
  window.addEventListener('DOMContentLoaded', e => {
    ...
    const ac = new AbortController();
    const input = document.querySelector('input[autocomplete="one-time-code"]');
    navigator.credentials.get({
      otp: { transport:['sms'] },
      signal: ac.signal
    }).then(otp => {
      input.value = otp.code;
    }).catch(err => { /* ... */ });
  });
}
```

Learn more about how to implement webOTP: [https://developer.chrome.com/articles/web-otp/](https://developer.chrome.com/articles/web-otp/).

Support for webOTP has been added since Chrome 93, Edge 93. Firefox and Safari have no support, meaning that webOTP is not readily available yet.

### File System Access

[Dave Bitter](https://techhub.iodigital.com/authors/dave-bitter) has written an excellent article about [File system access](https://techhub.iodigital.com/articles/the-file-system-access-api)

## Features in the pipeline

> Some of the features that are being developed or considered at the moment

Reviewer: ```suggestion
Some of the features that are being developed or considered at the moment:
```

- Changing system settings API
- Geofencing
- Splash screen
- Remote desktop control
- Call dialer/answering/control

## ...but what about the other one?

> Project Fugu can offer a great experience, but only if the browser supports these APIs. Safari (and thus iOS and iPad OS) are not actively part of project fugu and have been slow with implementing APIs that extend the functionality of the web, like service worker and push notifications. Perhaps that will change, but it might not. Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. Having APIs fully specced and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.

Reviewer: ```suggestion
Project Fugu can offer a great experience, but only if the browser supports these APIs. Safari (and thus iOS and iPad OS) are not actively part of Project Fugu and have been slow with implementing APIs that extend the functionality of the web. Eg. support for service workers and push notifications. 

Perhaps that will change, but it might not. Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. 

Having APIs fully specified and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.
```

> Project Fugu can offer a great experience, but only if the browser supports these APIs. Safari (and thus iOS and iPad OS) are not actively part of project fugu and have been slow with implementing APIs that extend the functionality of the web, like service worker and push notifications. Perhaps that will change, but it might not. Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. Having APIs fully specced and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.

Reviewer: IMHO, I think the phrase `slow with implementing APIs that extend the functionality of the web`

Although SW and Push support have been a long time coming, there are clear indications that the Safari team is moving the web platform forward. Look at https://wpt.fyi/interop-2023

Maybe rephrase this sentence with more information? I think you are aware what forces are currently in play, eg. EU regulations.

For example:

> If Safari and Apple will ever actively contribute to Project Fugu, still remains an open question. On the one hand, the fact that Apple does not allow other browsers engines from their mobile platforms, is one of the main reasons why web apps are still considered less than a native app. On the other, EU regulators have noticed this as well and are actively in debate if this 'browser ban' is considered to be legal. For more information concerning this specific topic, check out https://open-web-advocacy.org/

## Closing thoughts

Cross-platform development is hard. Features and capabilities of an application require you to think about how it needs to be implemented in different platforms like Android, iOS, Web, Windows, MacOS and Linux. If you, for example, need to access the address details of a contact that is stored on a device you need to implement a contact picker for all the different platforms you want to support or you need to come up with a completely custom implementation. Custom implementations are probably not what you want to do. It can become quite messy quite fast.

So what can you do?

Progressive Web Apps (PWAs) bring offline support and _app-like_ experiences to the web. Extending the PWA principle with APIs to include more native APIs while keeping the principles of the web (trust, privacy and security) is the core of Project Fugu 🐡.

Some of these APIs are readily available and should be considered when that feature is needed, like for example the async clipboard API. Other APIs though are only available in certain browsers, like the webOTP API so handle them with care and consult the online references.

> Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. Having APIs fully specced and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.

Reviewer: I'd like to refer to my previous comment about Safari and Apple.
